rust_snuba/rust_arroyo/src/processing/strategies/mod.rs

```diff
@@ -0,0 +1,173 @@
+        for (partition, offset) in other.positions {
+            if self.positions.contains_key(&partition) {
+                if self.positions[&partition] < offset {
+                    self.positions.insert(partition, offset);
+                }
+            } else {
+                self.positions.insert(partition, offset);
+            }
+        }
+
+        self
+    }
+}
+
+pub fn merge_commit_request(
+    value: Option<CommitRequest>,
+    other: Option<CommitRequest>,
+) -> Option<CommitRequest> {
+    match (value, other) {
+        (None, None) => None,
+        (Some(x), None) => Some(x),
+        (None, Some(y)) => Some(y),
+        (Some(x), Some(y)) => Some(x.merge(y)),
+    }
+}
+
+/// A processing strategy defines how a stream processor processes messages
+/// during the course of a single assignment. The processor is instantiated
+/// when the assignment is received, and closed when the assignment is
+/// revoked.
+///
+/// This interface is intentionally not prescriptive, and affords a
+/// significant degree of flexibility for the various implementations.
+pub trait ProcessingStrategy<TPayload: Clone>: Send + Sync {
+    /// Poll the processor to check on the status of asynchronous tasks or
+    /// perform other scheduled work.
+    ///
+    /// This method is called on each consumer loop iteration, so this method
+    /// should not be used to perform work that may block for a significant
+    /// amount of time and block the progress of the consumer or exceed the
+    /// consumer poll interval timeout.
+    ///
+    /// This method may raise exceptions that were thrown by asynchronous
+    /// tasks since the previous call to ``poll``.
+    fn poll(&mut self) -> Result<Option<CommitRequest>, InvalidMessage>;
+
+    /// Submit a message for processing.
+    ///
+    /// Messages may be processed synchronously or asynchronously, depending
+    /// on the implementation of the processing strategy. Callers of this
+    /// method should not assume that this method returning successfully
+    /// implies that the message was successfully processed.
+    ///
+    /// If the processing strategy is unable to accept a message (due to it
+    /// being at or over capacity, for example), this method will raise a
+    /// ``MessageRejected`` exception.
+    fn submit(&mut self, message: Message<TPayload>) -> Result<(), SubmitError<TPayload>>;
+
+    /// Close this instance. No more messages should be accepted by the
+    /// instance after this method has been called.
+    ///
+    /// This method should not block. Once this strategy instance has
+    /// finished processing (or discarded) all messages that were submitted
+    /// prior to this method being called, the strategy should commit its
+    /// partition offsets and release any resources that will no longer be
+    /// used (threads, processes, sockets, files, etc.)
+    fn close(&mut self);
```

Comment:
No, it's more like Drop right now. Though `join` can be passed a timeout. 

The reason I mentioned "right now" in this answer is that it's likely we will want to change this API in the future to reuse strategies across rebalances (and thus flushing them instead of fully closing/joining them every time). The main motivation for this is to reduce the amount of work done in rebalancing callbacks (this is especially painful in strategies that are slow to initialize - like the python multiprocessing strategy).